inspirehep/modules/records/permissions.py

```diff
@@ -0,0 +1,105 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of INSPIRE.
+# Copyright (C) 2016 CERN.
+#
+# INSPIRE is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# INSPIRE is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
+#
+# In applying this licence, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+from __future__ import absolute_import, division, print_function
+
+from flask_security import current_user
+
+from invenio_access.models import ActionUsers
+from invenio_access.permissions import (DynamicPermission,
+                                        ParameterizedActionNeed)
+
+
+action_view_restricted_collection = ParameterizedActionNeed(
+    'view-restricted-collection', argument=None
+)
+
+
+def record_read_permission_factory(record=None):
+    """Record permission factory."""
+    return RecordPermission.create(record=record, action='read')
+
+
+class RecordPermission(object):
+    """Record permission.
+
+    - Read access given if collection not restricted.
+    - All other actions are denied for the moment.
+    """
+
+    read_actions = ['read']
+
+    def __init__(self, record, func, user):
+        """Initialize a file permission object."""
+        self.record = record
+        self.func = func
+        self.user = user or current_user
+
+    def can(self):
+        """Determine access."""
+        return self.func(self.user, self.record)
+
+    @classmethod
+    def create(cls, record, action, user=None):
+        """Create a record permission."""
+        if action in cls.read_actions:
+            return cls(record, has_read_permission, user)
+        else:
+            return cls(record, deny, user)
+
+
+def has_read_permission(user, record):
+    """Check if user has read access to the record."""
+    user_roles = [r.name for r in current_user.roles]
+    if 'superuser' in user_roles:
+        return True
+
+    all_restricted_collections = set(
+        [
+            a.argument for a in ActionUsers.query.filter_by(
+                action='view-restricted-collection').all()
+        ]
+    )
```

Comment:
Shouldn't this be computed only once, for performance reason. We could cache it in Redis for a minute or so.